src/textconv.c

```diff
@@ -0,0 +1,611 @@
+    
+    if (git_rwlock_wrlock(&textconv_registry.lock) < 0) {
+        giterr_set(GITERR_OS, "failed to lock textconv registry");
+        return -1;
+    }
+    
+    if (!textconv_registry_find(NULL, name)) {
+        giterr_set(
+                   GITERR_TEXTCONV, "attempt to reregister existing textconv '%s'", name);
+        error = GIT_EEXISTS;
+        goto done;
+    }
+    
+    error = textconv_registry_insert(name, textconv);
+    
+done:
+    git_rwlock_wrunlock(&textconv_registry.lock);
+    return error;
+}
+
+int git_textconv_unregister(const char *name)
+{
+    size_t pos;
+    git_textconv_entry *entry;
+    int error = 0;
+    
+    assert(name);
+    
+    if (git_rwlock_wrlock(&textconv_registry.lock) < 0) {
+        giterr_set(GITERR_OS, "failed to lock textconv registry");
+        return -1;
+    }
+    
+    if ((entry = textconv_registry_lookup(&pos, name)) == NULL) {
+        giterr_set(GITERR_TEXTCONV, "cannot find textconv '%s' to unregister", name);
+        error = GIT_ENOTFOUND;
+        goto done;
+    }
+    
+    git_vector_remove(&textconv_registry.textconvs, pos);
+    
+    if (entry->initialized && entry->textconv && entry->textconv->shutdown) {
+        entry->textconv->shutdown(entry->textconv);
+        entry->initialized = false;
+    }
+    
+    git__free(entry->textconv_name);
+    git__free(entry);
+    
+done:
+    git_rwlock_wrunlock(&textconv_registry.lock);
+    return error;
+}
+
+static int textconv_initialize(git_textconv_entry *entry)
+{
+    int error = 0;
+    
+    if (!entry->initialized && entry->textconv && entry->textconv->initialize) {
+        if ((error = entry->textconv->initialize(entry->textconv)) < 0)
+            return error;
+    }
+    
+    entry->initialized = true;
+    return 0;
+}
+
+git_textconv *git_textconv_lookup(const char *name)
+{
+    size_t pos;
+    git_textconv_entry *entry;
+    git_textconv *textconv = NULL;
+    
+    if (git_rwlock_rdlock(&textconv_registry.lock) < 0) {
+        giterr_set(GITERR_OS, "failed to lock textconv registry");
+        return NULL;
+    }
+    
+    if ((entry = textconv_registry_lookup(&pos, name)) == NULL ||
+        (!entry->initialized && textconv_initialize(entry) < 0))
+        goto done;
+    
+    textconv = entry->textconv;
+    
+done:
+    git_rwlock_rdunlock(&textconv_registry.lock);
+    return textconv;
+}
+
+void git_textconv_free(git_textconv *textconv)
+{
+    git__free(textconv);
+}
+
+
+
+/** Return any textconv applicable to the source */
+static int git_textconv_check_attributes(
+                                        const char **out,
+                                        git_repository *repo,
+                                        const char *path)
+{
+    int error;
+    
+    error = git_attr_get(out, repo, 0, path, GIT_TEXTCONV_ATTR);
+    
+    /* if no values were found there is no textconv, return passthrough */
+    if (error == GIT_ENOTFOUND) {
+        return GIT_PASSTHROUGH;
```

Comment:
`GIT_PASSTHROUGH` is supposed to be a special return code from our callbacks only (meaning "there was a callback, but it want to be ignored"), so that would be somewhat unexpected externally. You might want to keep that `GIT_ENOTFOUND` instead.